apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/helper/DubboTraceHelper.java

```diff
@@ -0,0 +1,162 @@
+        String version = apiInfo.getVersion();
+        if (version != null && version.length() > 0) {
+            requestName += " version=" + version;
+        }
+
+        return requestName;
+    }
+
+    @VisibleForAdvice
+    public static Span createConsumerSpan(DubboApiInfo apiInfo, InetSocketAddress remoteAddress) {
+        TraceContextHolder<?> traceContext = DubboTraceHelper.tracer.getActive();
+        if (traceContext == null) {
+            return null;
+        }
+        Span span = traceContext.createExitSpan();
+        if (span == null) {
+            return null;
+        }
+
+        span.withType(EXTERNAL_TYPE)
+            .withSubtype(DUBBO_SUBTYPE)
+            .withName(buildDubboRequestName(apiInfo));
+        Destination destination = span.getContext().getDestination();
+        destination.withAddress(remoteAddress.getHostName()).withPort(remoteAddress.getPort());
+
+        Destination.Service service = destination.getService();
+        service.withType(EXTERNAL_TYPE).withResource(DUBBO_SUBTYPE).withName(DUBBO_SUBTYPE);
+
+        return span.activate();
+    }
+
+    @VisibleForAdvice
+    public static void fillTransaction(Transaction transaction, DubboApiInfo dubboApiInfo) {
+        transaction.withName(buildDubboRequestName(dubboApiInfo));
+        transaction.withType("dubbo");
+        transaction.activate();
+    }
+
+    public static boolean isBizException(Class<?> interfaceClass, Class<?> exp) {
+        String apiJarFile = getJarFile(interfaceClass);
+        if (apiJarFile == null) {
+            return false;
+        }
+        return apiJarFile.equals(getJarFile(exp));
+    }
+
+    public static String getJarFile(Class<?> clazz) {
+        if (clazz == null) {
+            return null;
+        }
+        ProtectionDomain domain = clazz.getProtectionDomain();
+        if (domain == null) {
+            return null;
+        }
+        CodeSource source = domain.getCodeSource();
+        if (source == null) {
+            return null;
+        }
+        URL location = source.getLocation();
+        if (location == null) {
+            return null;
+        }
+        return location.getFile();
+    }
+
+    public static void doCapture(Object[] args, Throwable t, Object returnValue) {
+        Transaction transaction = tracer.currentTransaction();
+        if (transaction == null) {
+            return;
+        }
+        boolean hasError = t != null;
+        CoreConfiguration coreConfig = tracer.getConfig(CoreConfiguration.class);
+        CoreConfiguration.EventType captureBody = coreConfig.getCaptureBody();
+        if (CoreConfiguration.EventType.OFF.equals(captureBody) ||
+            (CoreConfiguration.EventType.ERRORS.equals(captureBody) && !hasError)) {
+            return;
+        }
+
+        captureArgs(transaction, args);
+        if (t != null) {
+            transaction.addCustomContext("throw", t.toString());
```

Comment:
Would `transaction.captureEception(t);` work?

Reply:
some business exceptions are not expected to be capture as an exception, in these case,  i want to capture their as a result, and store it in custom context